src/main/scala/uk/gov/nationalarchives/tdr/validation/schema/MetadataValidationJsonSchema.scala

```diff
@@ -20,52 +16,50 @@ object MetadataValidationJsonSchema {
   def validate(metadata: List[FileRow]): Map[String, List[Error]] = {
     val convertedFileRows: Seq[ObjectMetadata] = metadata.map(fileRow => ObjectMetadata(fileRow.fileName, fileRow.metadata.toSet))
-    validate(BASE_SCHEMA, convertedFileRows.toSet)
+    val validationProgram = for {
+      jsonData <- IO(convertedFileRows.map(objectMetadata => mapToJson(objectMetadata)))
+      validationErrors <- IO(jsonData.map(jsonData => validateWithSchema(BASE_SCHEMA)(jsonData)))
+      closureValidationErrors <- IO(jsonData.map(jsonData => validateWithSchema(CLOSURE_SCHEMA)(jsonData)))
+      errors <- convertSchemaValidatorError(validationErrors ++ closureValidationErrors)
+    } yield errors.toMap
+
+    validationProgram.unsafeRunSync()
   }
 
   /*
    Validate against specified schema
    */
   def validate(schemaDefinition: JsonSchemaDefinition, metadata: Set[ObjectMetadata]): Map[String, List[Error]] = {
     val validationProgram = for {
-      validationErrors <- schemaValidation(schemaDefinition, metadata)
-      errors <- convertSchemaValidatorError(validationErrors)
+      jsonData <- IO(metadata.map(objectMetadata => mapToJson(objectMetadata)))
+      validationErrors <- IO(jsonData.map(jsonData => validateWithSchema(schemaDefinition)(jsonData)))
+      errors <- convertSchemaValidatorError(validationErrors.toSeq)
     } yield errors.toMap
 
     validationProgram.unsafeRunSync()
   }
 
-  private def schemaValidation(schemaDefinition: JsonSchemaDefinition, metadata: Set[ObjectMetadata]): IO[Seq[ValidationErrors]] = {
-    IO.fromFuture(
-      IO(
-        Source(metadata)
-          .map(mapToJson)
-          .map(validateWithSchema(schemaDefinition))
-          .runWith(Sink.seq[ValidationErrors])
-      )
-    )
-  }
-
-  private def validateWithSchema(schemaDefinition: JsonSchemaDefinition) = { (jsonData: JsonData) =>
+  private def validateWithSchema(schemaDefinition: JsonSchemaDefinition): JsonData => ValidationErrors = { (jsonData: JsonData) =>
     schemaDefinition match {
       case BASE_SCHEMA =>
         val errors = JsonSchemaValidators.validateJson(schemaDefinition, jsonData.json)
         ValidationErrors(BASE_SCHEMA_VALIDATION, jsonData.identifier, errors)
+      case CLOSURE_SCHEMA =>
+        val errors = JsonSchemaValidators.validateJson(schemaDefinition, jsonData.json)
+        ValidationErrors(CLOSURE_SCHEMA_VALIDATION, jsonData.identifier, errors)
     }
   }
 
   /*
    What we want to use for the errors has yet to be defined
    */
   private def convertSchemaValidatorError(errors: Seq[MetadataValidationJsonSchema.ValidationErrors]): IO[Seq[(String, List[Error])]] = {
-    IO(errors.map(error => error.identifier -> error.errors.map(error => Error(error.getInstanceLocation.getName(0), error.getMessageKey)).toList))
+    IO(errors.map(error => error.identifier -> error.errors.map(e => Error(Option(e.getProperty).getOrElse(e.getInstanceLocation.getName(0)), e.getMessageKey)).toList))
```

Comment:
As the mapping has grown to it be pulled out to a separate function def
`IO(errors.map(error => error.identifier -> error.errors.map(convertValidationMessage).toList))`